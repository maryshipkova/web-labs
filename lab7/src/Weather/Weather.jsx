import * as React from 'react';
import './Weather.scss';

const getData = (city, coordinates) => {
    const fetchUrl = city ? `q=${city}` : `lat=${coordinates.lat}&lon=${coordinates.lon}`;

    return fetch(`http://api.openweathermap.org/data/2.5/forecast?${fetchUrl}&APPID=62462c93c650ac75e405b900f2457d73`)
        .then(data => data.json())
};

export const Weather = (props) => {
    const [weatherInfo, updateWeatherInfo] = React.useState({});
    const [loading, isLoading] = React.useState(false);

    const getCityData = (city) => {
        return getData(city, coordinates)
            .then(data => {
                const currWeather = data.list[0];
                updateWeatherInfo({
                    cityName: data.city.name,
                    coord: [data.city.coord.lat, data.city.coord.lon],
                    humidity: `${currWeather.main.humidity} %`,
                    pressure: `${currWeather.main.pressure} hPa`,
                    wind: `${currWeather.wind.speed} meter/sec`,
                    weather: currWeather.weather[0].description,
                    temperature: `${parseInt((currWeather.main.temp - 272) + '')} C`,
                    icon: currWeather.weather[0].icon
                });
                isLoading(false);
            })
            .catch(e => {
                isLoading(false);
                return updateWeatherInfo({city, error: '404 city not found'});
            })
    };

    const {city, coordinates, main, onRemove} = props;

    React.useEffect(() => {
        if (city || coordinates) {
            isLoading(true);
            getCityData(city, coordinates);
        }
    }, [city, coordinates]);


    if(loading) return <div>loading</div>;
    if (weatherInfo.error){
        return (
            <div>
                {city}: {weatherInfo.error}
                <button className={'Weather-Remove'} onClick={() => onRemove(city)}>&Chi;</button>
            </div>);
    }

    const {cityName, temperature, wind, weather, pressure, humidity, coord, icon} = weatherInfo;

    return (
        <div className={`Weather${main ? ' Weather_main' : ''}`}>
            {weatherInfo && weatherInfo.cityName ?
                <React.Fragment>
                    <div className="Weather-General">
                        <h1 className="Weather-City">{cityName}</h1>
                        <div className="Weather-MainInfo">
                            <div className="Weather-Icon">
                                <img src={`http://openweathermap.org/img/wn/${icon}@2x.png`} alt={icon}/>
                            </div>
                            <h2 className="Weather-Temperature">{temperature}</h2>
                        </div>
                        {!main && <button className={'Weather-Button'} onClick={() => onRemove(city)}>&Chi;</button>}
                    </div>
                    <ul className="Weather-List">
                        <li className="Weather-Item">Ветер: {wind}</li>
                        <li className="Weather-Item">Облачность: {weather}</li>
                        <li className="Weather-Item">Давление: {pressure}</li>
                        <li className="Weather-Item">Влажность: {humidity}</li>
                        <li className="Weather-Item">Координаты: {coord && `${coord[0]} ${coord[1]}`}</li>
                    </ul>
                </React.Fragment> : null}
        </div>
    )
};
