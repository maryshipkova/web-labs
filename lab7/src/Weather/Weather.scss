@import '../constants.scss';

.Weather {
  &:not(:last-child){
    margin-bottom: 20px;
  }

  &-General {
    display: flex;
    align-items: center;
    font-size: 16px;
  }

  &-City {
    margin: 0;
  }

  &-MainInfo {
    display: flex;
    align-items: center;
    margin-left: 10px;
  }

  &-Temperature {
    margin: 0;
    margin-left: 10px;
  }

  &-List {
    padding: 0;
  }

  &-Item {
    padding: 5px 10px;
    margin-bottom: 5px;

    background-color: $color-gray;
    color: white;
    list-style: none;
  }
}

.Weather_main.Weather {
  display: flex;
  margin-top: 20px;

  .Weather-General {
    flex-direction: column;
    align-items: flex-start;
  }

  .Weather-MainInfo{
    margin-top: 20px;

  }

  .Weather-List {
    padding: 0;
    margin-left: 90px;
  }

  .Weather-Temperature {
    font-size: 50px;
  }
}
