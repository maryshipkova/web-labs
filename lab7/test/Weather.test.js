import React from 'react';
import renderer from 'react-test-renderer';
import {Weather} from "../src/Weather/Weather";
import thunk from "redux-thunk";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";

describe()
it('Weather', () => {
    const initialState={cities: ["Moscow","1234","12345"]};
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    const weatherInfo = {
        cityName: "Moscow",
            coord: (2) [55.7507, 37.6177],
        humidity: "77 %",
        icon: "04n",
        pressure: "1030 hPa",
        temperature: "3 C",
        weather: "overcast clouds",
        wind: "4.2 meter/sec"
    };
    const tree = renderer
        .create(
            <Provider store={store}>
                <Weather weatherInfo={weatherInfo}/>
            </Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('Weather Main', () => {
    const initialState={cities: ["Moscow","1234","12345"]};
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    const weatherInfo = {error:'test'};
    const tree = renderer
        .create(
            <Provider store={store}>
                <Weather main weatherInfo={weatherInfo}/>
            </Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('Weather 404', () => {
    const initialState={cities: ["Moscow","1234","12345"]};
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    const weatherInfo = {
        cityName: "Moscow",
        coord: (2) [55.7507, 37.6177],
        humidity: "77 %",
        icon: "04n",
        pressure: "1030 hPa",
        temperature: "3 C",
        weather: "overcast clouds",
        wind: "4.2 meter/sec"
    };
    const tree = renderer
        .create(
            <Provider store={store}>
                <Weather weatherInfo={weatherInfo}/>
            </Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('Weather w/o data', () => {
    const initialState={cities: ["Moscow","1234","12345"]};
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    const weatherInfo = {
        cityName: "Moscow",
        coord: (2) [55.7507, 37.6177],
        humidity: "77 %",
        icon: "04n",
        pressure: "1030 hPa",
        temperature: "3 C",
        weather: "overcast clouds",
        wind: "4.2 meter/sec"
    };
    const tree = renderer
        .create(
            <Provider store={store}>
                <Weather weatherInfo={weatherInfo}/>
            </Provider>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

