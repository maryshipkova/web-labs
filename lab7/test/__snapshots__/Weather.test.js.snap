// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather 1`] = `
<div
  className="Weather"
>
  <div
    className="Weather-General"
  >
    <h1
      className="Weather-City"
    >
      Moscow
    </h1>
    <div
      className="Weather-MainInfo"
    >
      <div
        className="Weather-Icon"
      >
        <img
          alt="04n"
          src="http://openweathermap.org/img/wn/04n@2x.png"
        />
      </div>
      <h2
        className="Weather-Temperature"
      >
        3 C
      </h2>
    </div>
    <button
      className="Weather-Button"
      onClick={[Function]}
    >
      Χ
    </button>
  </div>
  <ul
    className="Weather-List"
  >
    <li
      className="Weather-Item"
    >
      Ветер: 
      4.2 meter/sec
    </li>
    <li
      className="Weather-Item"
    >
      Облачность: 
      overcast clouds
    </li>
    <li
      className="Weather-Item"
    >
      Давление: 
      1030 hPa
    </li>
    <li
      className="Weather-Item"
    >
      Влажность: 
      77 %
    </li>
    <li
      className="Weather-Item"
    >
      Координаты: 
    </li>
  </ul>
</div>
`;

exports[`Weather 2`] = `
<div>
  : 
  test
  <button
    className="Weather-Remove"
    onClick={[Function]}
  >
    Χ
  </button>
</div>
`;

exports[`Weather 404  1`] = `
<div>
  : 
  test
  <button
    className="Weather-Remove"
    onClick={[Function]}
  >
    Χ
  </button>
</div>
`;
